packages/apify/src/internals/charging.ts

```diff
@@ -0,0 +1,281 @@
+import log from '@apify/log';
+import { Dataset, KeyValueStore } from '@crawlee/core';
+import { ApifyClient } from 'apify-client';
+
+import { Configuration } from '../configuration.js';
+
+/**
+ * Handles pay-per-event charging.
+ *
+ * @internal
+ */
+export class ChargingManager {
+    public readonly LOCAL_CHARGING_LOG_DATASET_NAME = 'charging_log';
+    public readonly PLATFORM_CHARGING_LOG_DATASET_ID_KEY = 'CHARGING_LOG_DATASET_ID';
+
+    private maxTotalChargeUsd: number;
+    private isAtHome: boolean;
+    private actorRunId: string | undefined;
+    private pricingModel: string | undefined = undefined;
+    private purgeChargingLogDataset: boolean;
+    private notPpeWarningPrinted = false;
+
+    private pricingInfo: Record<string, {price: number; title: string}> = {};
+    private chargingState: Record<string, ChargingStateItem> | undefined = undefined;
+    private chargingLogDataset: Dataset | undefined;
+
+    private apifyClient: ApifyClient;
+
+    constructor(configuration: Configuration, apifyClient: ApifyClient) {
+        this.maxTotalChargeUsd = configuration.get('maxTotalChargeUsd') ?? Infinity;
+        this.isAtHome = configuration.get('isAtHome');
+        this.actorRunId = configuration.get('actorRunId');
+        this.purgeChargingLogDataset = configuration.get('purgeOnStart');
+
+        if (configuration.get('testPayPerEvent') === true) {
+            if (this.isAtHome) {
+                throw new Error('Using the ACTOR_TEST_PAY_PER_EVENT environment variable is only supported in a local development environment');
+            }
+
+            this.pricingModel = 'PAY_PER_EVENT';
+        }
+
+        this.apifyClient = apifyClient;
+    }
+
+    private get isPayPerEvent() {
+        return this.pricingModel === 'PAY_PER_EVENT';
+    }
+
+    /**
+     * Initialize the ChargingManager by loading pricing information and charging state via Apify API.
+     */
+    async init(): Promise<void> {
+        this.chargingState = {};
+
+        // Retrieve pricing information
+        if (this.isAtHome) {
```

Comment:
Q: what does exactly `isAtHome` mean? Would be good to document, it's non-obvious to me.

Reply:
The value depends on an environment variable set by the platform and it is used to distinguish local runs from runs on Apify. It's ubiquitous in the SDK code, so I don't see the need to explain it here.

Reply:
Note that we also have `Actor.isAtHome()`, so this shouldn't be anything new even for our users, not just for contributors.

https://docs.apify.com/sdk/js/reference/class/Actor#isAtHome